### PR TITLE
feat: Endpoint for 'sanctioned' order message

### PR DESCRIPTION
**Pull request by JadeyOlivier**
### Description
Added a view to coordinator for the AWS Eventbridge API destination to point to when receiving a Commercetools subscription message to sanction an order. 

### JIRA
https://2u-internal.atlassian.net/browse/SONIC-165

### CODE REVIEW

File: commerce_coordinator/apps/commercetools/urls.py [hunk 1]
```diff
@@ -4,10 +4,11 @@
 
 from django.urls import include, path
 
-from commerce_coordinator.apps.commercetools.views import OrderFulfillView
+from commerce_coordinator.apps.commercetools.views import OrderFulfillView, OrderSanctionedView
 
 app_name = 'commercetools'
 urlpatterns = [
     path('api-auth/', include('rest_framework.urls', namespace='rest_framework')),
     path('fulfill', OrderFulfillView.as_view(), name='fulfill'),
```
Comment (grmartin) [inline commerce_coordinator/apps/commercetools/urls.py:12]:
```suggestion
    # EventBridge / CloudWatch Endpoints
    path('fulfill', OrderFulfillView.as_view(), name='fulfill'),
```

File: commerce_coordinator/apps/commercetools/views.py [hunk 2]
```diff
@@ -82,3 +89,39 @@ def post(self, request):
         send_order_confirmation_email(lms_user_id, customer.email, canvas_entry_properties)
 
         return Response(status=status.HTTP_200_OK)
+
+class OrderSanctionedView(APIView):
+    """View to sanction an order and deactivate the lms user"""
+    permission_classes = [IsAdminUser]
+    authentication_classes = [] #TODO: Update this with OAuth authentication
```
Comment (grmartin) [inline commerce_coordinator/apps/commercetools/views.py]: If you make this an empty list there are no ayth classes... and thus it cant be hit... (given its permissions_classes requirements)
```suggestion
    # authentication_classes = [] # TODO: Update this with OAuth authentication
```

**Review by grmartin — APPROVED** [review]

File: commerce_coordinator/apps/commercetools/views.py [hunk 3]
```diff
@@ -82,3 +89,40 @@ def post(self, request):
...
+
+        message_details = OrderMessageInputSerializer(data=input_data)
+        message_details.is_valid(raise_exception=True)
+
+        client = CommercetoolsAPIClient()
+        order_id = message_details.data['order_id']
+        order = client.get_order_by_id(order_id)
+        customer = client.get_customer_by_id(order.customer_id)
+
+        if not (customer and order and is_edx_lms_order(order)):
```
Comment (grmartin) [inline commerce_coordinator/apps/commercetools/views.py:118]: One last thing, you need to read the status and endure its set to sanctioned.

**Review by grmartin — CHANGES_REQUESTED** [review]

**Review by grmartin — APPROVED** [review]
Pending decision on coverage, proposed to the team...

File: commerce_coordinator/apps/commercetools/views.py [hunk 4]
```diff
@@ -82,3 +90,51 @@ def post(self, request):
         send_order_confirmation_email(lms_user_id, customer.email, canvas_entry_properties)
 
         return Response(status=status.HTTP_200_OK)
+
+
+class OrderSanctionedView(APIView):
+    """View to sanction an order and deactivate the lms user"""
+    permission_classes = [IsAdminUser]
+
+    # authentication_classes = [] TODO: Update this with OAuth authentication
```
Comment (shafqatfarhan) [inline commerce_coordinator/apps/commercetools/views.py:99]: No need to explicitly specify authentication_classes in the endpoint as we have `JwtAuthentication` set as default in [REST_FRAMEWORK](https://github.com/edx/commerce-coordinator/blob/main/commerce_coordinator/settings/base.py#L277)

File: commerce_coordinator/apps/commercetools/views.py [hunk 5]
```diff
@@ -82,3 +90,51 @@ def post(self, request):
...
+
+        message_details = OrderMessageInputSerializer(data=input_data)
+        message_details.is_valid(raise_exception=True)
+
+        client = CommercetoolsAPIClient()
+        order_id = message_details.data['order_id']
+        order = client.get_order_by_id(order_id)
+        order_workflow_state = order.state
+
+        customer = client.get_customer_by_id(order.customer_id)
```
Comment (shafqatfarhan) [inline commerce_coordinator/apps/commercetools/views.py:120]: Do we need to catch any API response that raises an error and add an exception log?

File: commerce_coordinator/apps/commercetools/views.py [hunk 6]
```diff
@@ -34,12 +44,18 @@ def post(self, request):
...
         order_id = message_details.data['order_id']
-        order = client.get_order_by_id(order_id)
+
+        try:
+            order = client.get_order_by_id(order_id)
+        except CommercetoolsError as _:  # pragma no cover
+            logger.error("Order not found: %s", order_id)
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
         customer = client.get_customer_by_id(order.customer_id)
```
Comment (shafqatfarhan) [inline commerce_coordinator/apps/commercetools/views.py:59]: Suggestion:
- Please move `get_customer_by_id` into try/except block as well
- Change the logger function to exception as we need to be alerted on just exceptions
- Make exception message more meaningful and append API error as well

```
try:
    order = client.get_order_by_id(order_id)
    customer = client.get_customer_by_id(order.customer_id)
except CommercetoolsError as exc:
    logger.exception(f'Error fetching CT Customer or Order with ID "{order_id}". Error: {exc}')
    return Response(status=status.HTTP_404_NOT_FOUND)
```

File: commerce_coordinator/apps/commercetools/views.py [hunk 7]
```diff
@@ -82,3 +98,56 @@ def post(self, request):
...
+
+        try:
+            order = client.get_order_by_id(order_id)
+        except CommercetoolsError as _:  # pragma no cover
+            logger.error("Order not found: %s", order_id)
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        order_workflow_state = order.state.obj.key
+
+        customer = client.get_customer_by_id(order.customer_id)
```
Comment (shafqatfarhan) [inline commerce_coordinator/apps/commercetools/views.py:133]: Suggestion:
- Please move `get_customer_by_id` into try/except block as well
- Change the logger function to exception as we need to be alerted on just exceptions
- Make exception message more meaningful and append API error as well

```
try:
    order = client.get_order_by_id(order_id)
    customer = client.get_customer_by_id(order.customer_id)
except CommercetoolsError as exc:
    logger.exception(f'Error fetching CT Customer or Order with ID "{order_id}". Error: {exc}')
    return Response(status=status.HTTP_404_NOT_FOUND)
```

**Pull request merged**